src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.AddProperty.cs

```diff
@@ -209,18 +210,26 @@ internal JsonPropertyInfo CreateRootObject(JsonSerializerOptions options)
                 options: options);
         }
 
-        internal JsonPropertyInfo CreatePolymorphicProperty(JsonPropertyInfo property, Type runtimePropertyType, JsonSerializerOptions options)
+        internal JsonPropertyInfo GetOrAddPolymorphicProperty(JsonPropertyInfo property, Type runtimePropertyType, JsonSerializerOptions options)
         {
-            JsonPropertyInfo runtimeProperty = CreateProperty(
-                property.DeclaredPropertyType, runtimePropertyType,
-                property.ImplementedPropertyType,
-                property.PropertyInfo,
-                parentClassType: Type,
-                converter: null,
-                options: options);
-            property.CopyRuntimeSettingsTo(runtimeProperty);
+            static JsonPropertyInfo CreateRuntimeProperty(Type runtimePropertyType, (JsonPropertyInfo property, JsonSerializerOptions options, Type classType) arg)
+            {
+                JsonPropertyInfo runtimeProperty = CreateProperty(
+                    arg.property.DeclaredPropertyType, runtimePropertyType,
+                    arg.property.ImplementedPropertyType,
+                    arg.property.PropertyInfo,
+                    parentClassType: arg.classType,
+                    converter: null,
+                    options: arg.options);
+
+                arg.property.CopyRuntimeSettingsTo(runtimeProperty);
+
+                return runtimeProperty;
+            }
+
+            var cache = LazyInitializer.EnsureInitialized(ref property.RuntimePropertyCache, () => new ConcurrentDictionary<Type, JsonPropertyInfo>());
 
-            return runtimeProperty;
+            return cache.GetOrAdd(runtimePropertyType, (type, arg) => CreateRuntimeProperty(type, arg), (property, options, Type));
```

Comment:
@steveharter, unrelated to this PR, but here is some code cleanup we should consider (separately):
Using `Type` as a property name makes code like this quite hard to parse/reason about. I was confused where `Type` is coming from and how we were passing in `System.Type` (i.e. a type) as a method parameter.
https://github.com/dotnet/corefx/blob/437d9bc76489ff18bd9ce04f287e07591f6bc38c/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L77

Reply:
Is "ClassType" better? (somewhat redundant) Or just doc?

Reply:
We already have a `ClassType` property, so that wouldn't work.

https://github.com/dotnet/corefx/blob/437d9bc76489ff18bd9ce04f287e07591f6bc38c/src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L42